File: 0994-rotting-oranges/0994-rotting-oranges.cpp
// TC- O(m*n)
// SC- O(m*n)

class Solution {
    int m,n;
    int fresh_oranges = 0;
    vector<int> directions = {-1,0,1,0,-1};

public:
    int orangesRotting(vector<vector<int>>& grid) {
        m = grid.size();
        n = grid[0].size();
        queue<pair<int,int>> q;

        // count number of fresh oranges to check there is none remaining at the end and add rotting orange to queue
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j]==1)
                    fresh_oranges++;
                if(grid[i][j]==2)
                    q.push({i,j});
            }
        }
        
        int minutes=0;
        while(!q.empty() and fresh_oranges>0){
            int qsize = q.size();
            // to check if there is atleast one fresh orange converted to rotten. what if all are empty cells surrounding a rotten cell
            bool rotted=false;
            for(int i=0; i<qsize; i++){
                pair<int,int> curr=q.front();
                q.pop();
                for(int d=0; d<directions.size()-1; d++){
                    int new_r = curr.first+directions[d];
                    int new_c = curr.second+directions[d+1];
                    if(new_r>=0 and new_r<m and new_c>=0 and new_c<n and grid[new_r][new_c]==1){
                        // make cell rotten
                        grid[new_r][new_c] = 2;
                        fresh_oranges--;
                        q.push({new_r, new_c});
                        rotted=true;
                    }
                }
            }
            if(rotted)
                minutes++;
        }

        return fresh_oranges>0? -1:minutes;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.py
# TC- O(mn)
# SC- O(mn)

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        
        m, n = len(grid), len(grid[0])
        dirs = [-1, 0, 1, 0, -1]
        fresh = 0
        q = deque()

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh += 1
                if grid[i][j] == 2:
                    q.append((i, j))
        # print(fresh)

        min_time = 0
        while q:
            q_size = len(q)
            anything_rotted = False
            for i in range(q_size):
                r, c = q.popleft()
                grid[r][c] = 0 #mark as visited

                for d in range(len(dirs)-1):
                    next_r, next_c = r+dirs[d], c+dirs[d+1]
                    if 0<=next_r<m and 0<=next_c<n and grid[next_r][next_c]==1:
                        fresh -= 1
                        grid[next_r][next_c] = 0
                        anything_rotted = True
                        q.append((next_r, next_c))
            if anything_rotted:
                min_time += 1

        # print(fresh)
        if fresh!=0:
            return -1
        
        return min_time
                